src/Runner.Worker/ActionManager.cs

```diff
@@ -596,73 +597,89 @@ private async Task<bool> RepoExistsAsync(IExecutionContext executionContext, Pip
+            var actionDownloadInfos = default(WebApi.ActionDownloadInfoCollection);
+            for (var attempt = 1; attempt <= 3; attempt++)
+            {
+                try
                 {
-                    continue;
+                    actionDownloadInfos = await jobServer.ResolveActionDownloadInfoAsync(executionContext.Plan.ScopeIdentifier, executionContext.Plan.PlanType, executionContext.Plan.PlanId, new WebApi.ActionReferenceList { Actions = actionReferences }, executionContext.CancellationToken);
+                    break;
                 }
+                catch (Exception ex) when (attempt < 3)
+                {
+                    executionContext.Output(ex.Message);
+                    executionContext.Debug(ex.ToString());
+                    if (String.IsNullOrEmpty(Environment.GetEnvironmentVariable("_GITHUB_ACTION_DOWNLOAD_NO_BACKOFF")))
+                    {
+                        var backoff = BackoffTimerHelper.GetRandomBackoff(TimeSpan.FromSeconds(10), TimeSpan.FromSeconds(30));
+                        executionContext.Output($"Retrying in {backoff.TotalSeconds} seconds");
+                        await Task.Delay(backoff);
+                    }
+                }
+            }
 
-                var repositoryReference = action.Reference as Pipelines.RepositoryPathReference;
-                ArgUtil.NotNull(repositoryReference, nameof(repositoryReference));
+            ArgUtil.NotNull(actionDownloadInfos, nameof(actionDownloadInfos));
+            ArgUtil.NotNull(actionDownloadInfos.Actions, nameof(actionDownloadInfos.Actions));
+            var apiUrl = GetApiUrl(executionContext);
+            var defaultAccessToken = executionContext.GetGitHubContext("token");
+            var configurationStore = HostContext.GetService<IConfigurationStore>();
+            var runnerSettings = configurationStore.GetSettings();
 
-                var downloadInfo = default(ActionDownloadInfo);
+            foreach (var actionDownloadInfo in actionDownloadInfos.Actions.Values)
+            {
+                // Add secret
+                HostContext.SecretMasker.AddValue(actionDownloadInfo.Authentication?.Token);
 
+                // Temporary code: Fix token and download URL
                 if (runnerSettings.IsHostedServer)
                 {
-                    downloadInfo = new ActionDownloadInfo
-                    {
-                        NameWithOwner = repositoryReference.Name,
-                        Ref = repositoryReference.Ref,
-                        ArchiveLink = BuildLinkToActionArchive(apiUrl, repositoryReference.Name, repositoryReference.Ref),
-                        Authorization = authorization,
-                    };
+                    actionDownloadInfo.Authentication = new WebApi.ActionDownloadAuthentication { Token = defaultAccessToken };
+                    actionDownloadInfo.TarballUrl = actionDownloadInfo.TarballUrl.Replace("<GITHUB_API_URL>", apiUrl);
```

Comment:
are we still need `<GITHUB_API_URL>`? will the service return fully resolved URL?

Reply:
temporary, will delete soon